UI: Support filtering by span content / type / name

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/18683?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18683/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18683/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18683/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

<!-- Please fill in changes proposed in this PR. -->

### How is this PR tested?

- [] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

Support filtering traces by span content, name, and type in the MLflow UI

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses -->

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename Column label to Field in filter UI

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix operator to default to available operator when column changes

When selecting a column, set operator to the first available operator
for that column instead of always defaulting to EQUALS. This ensures
span content filter shows CONTAINS instead of 'Select'.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove SPAN_ATTRIBUTES_COLUMN_ID constant

Remove unused SPAN_ATTRIBUTES_COLUMN_ID constant and its import
statements, as well as the associated case statement in the filter
logic.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

daniellok-db commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/hooks/useMlflowTraces.test.tsx:
```diff
@@ -308,7 +315,7 @@ describe('useSearchMlflowTraces', () => {
             {
               column: 'custom_metadata:mlflow.trace.session',
-              operator: HiddenFilterOperator.ILIKE,
+              operator: 'ILIKE' as any,
```
same as above can also just use `CONTAINS` directly i think

dbczumar replied:
Totally - fixed! Replaced with `CONTAINS`

daniellok-db commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/hooks/useMlflowTraces.tsx:
```diff
@@ -640,6 +644,32 @@ const createMlflowSearchFilter = (
           break;
+        case SPAN_NAME_COLUMN_ID:
+          if (networkFilter.operator === '=') {
+            filter.push(`span.name ILIKE '${networkFilter.value}'`);
```
should this be `=`? or it's intentionally ILIKE for usability?

```suggestion
            filter.push(`span.name = '${networkFilter.value}'`);
```

daniellok-db reviewed (approved):
Overall LGTM once comments are addressed. However one caveat is that for databricks v4 experiments, the search bar already searches span inputs via API so there might be some UX inconsistency there as we have two ways to do the same / similar things. We were planning to port that over to work in OSS in the next UI sync.

daniellok-db commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/components/filters/TableFilterItem.tsx:
```diff
@@ -42,6 +46,42 @@ const FILTERABLE_INFO_COLUMNS = [
+    return true;
+  }
+  return false;
+};
```
commented below, i think i see what this was for: https://github.com/mlflow/mlflow/pull/18683#discussion_r2493679920

dbczumar committed:
Address PR review comments from @daniellok-db

- Replace ILIKE operator with CONTAINS in filters
- Remove supportsAllOperators function
- Refactor operator selector with isOperatorSelectorDisabled
- Update TableFilter operator type to use FilterOperator only
- Add CONTAINS to ILIKE conversion for custom metadata filters
- Update test expectations for CONTAINS operator

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/components/filters/TableFilterItem.tsx:
```diff
@@ -200,19 +248,18 @@ export const TableFilterItem = ({
-            value={column === '' || column === EXECUTION_DURATION_COLUMN_ID ? operator : '='}
-            disabled={column !== '' && column !== EXECUTION_DURATION_COLUMN_ID}
+            value={column === '' || supportsAllOperators(column, key) ? operator : getAvailableOperators(column, key)[0]}
+            disabled={column !== '' && getAvailableOperators(column, key).length === 1}
```
Makes sense! Fixed!

dbczumar committed:
Add inline comments documenting ILIKE usage for span filtering

Clarify that ILIKE is intentionally used instead of = for span.name and
span.type equals operators to provide case-insensitive matching for
better user experience.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/types.ts:
```diff
@@ -168,7 +168,7 @@ export type TableFilter = {
   // Should be defined if a column group is used.
   key?: string;
-  operator: FilterOperator | HiddenFilterOperator;
+  operator: FilterOperator | string;
```
Totally - fixed!

dbczumar commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/hooks/useMlflowTraces.tsx:
```diff
@@ -640,6 +644,32 @@ const createMlflowSearchFilter = (
+          break;
         default:
           if (networkFilter.column.startsWith(CUSTOM_METADATA_COLUMN_ID)) {
             const columnKey = `request_metadata.${getCustomMetadataKeyFromColumnId(networkFilter.column)}`;
```
Great catch! Fixed!

dbczumar committed:
Add missing FilterOperator import in useMlflowTraces.tsx

Import FilterOperator enum to fix TypeScript compilation error when
checking for CONTAINS operator in custom metadata filter conversion.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update i18n keys after changing filter label from Column to Field

Signed-off-by: dbczumar <corey.zumar@databricks.com>
